pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -526,6 +571,26 @@ void connectionOpened(final ClientCnx cnx) {
     }
 
     /**
+     * Clear the internal receiver queue and returns the message id of what was the 1st message in the queue that was
+     * not seen by the application
+     */
+    private MessageIdImpl clearReceiverQueue() {
+        List<Message> currentMessageQueue = new ArrayList<>(incomingMessages.size());
+        incomingMessages.drainTo(currentMessageQueue);
+        if (!currentMessageQueue.isEmpty()) {
+            return (MessageIdImpl) currentMessageQueue.get(0).getMessageId();
+        } else if (lastDequeuedMessage != null) {
```

Comment:
@rdhabalia  Apart from the wrong logic here (delete pos vs read pos). The tricky problem is to tell the broker the exact message from where we want to resume reading after reconnection. 

If we do have messages in the queue, we cannot rely on the `lastDequeuedMessage`, because another thread might be dequeueing at the same time. So we could see `msg1` there, while the application already has dequeued `msg2`. At that point the `msg2` will get duplicated after reconnection. 

By atomically draining the queue on the side, we can establish which was the last message "seen" by the application. Checking the queue size and draining must be done atomically.